File: Get_CAMD.R
rm(list=ls())

library(httr2)
library(tidyverse)
library(jsonlite)


setwd("C:/Users/laure/Documents/Gen_Plant_Emissions/")

source("Gen_Plant_Emissions/API_Key.R")

# facilities = request("https://api.epa.gov/easey/facilities-mgmt/facilities")%>%
#   req_headers(`x-api-key`=API_Key,
#               `accept`= "application/json")%>%
#   req_perform()
# 
# facilities = fromJSON(rawToChar(facilities$body))

facilities_attributes_data=NULL
for(y in 2015:2020){
  url=paste("https://api.epa.gov/easey/facilities-mgmt/facilities/attributes?page=1&perPage=100&year=", y, sep="")
  facilities_attributes = request(url)%>%
    req_headers(`x-api-key`=API_Key,
                `accept`= "application/json")%>%
    req_perform()
  
  num_pages = ceiling(as.numeric(facilities_attributes$headers$`X-Total-Count`)/100)
  
  facilities_attributes = fromJSON(rawToChar(facilities_attributes$body))
  facilities_attributes_data=rbind(facilities_attributes_data, facilities_attributes)
  for(p in 2:num_pages){
    url=paste("https://api.epa.gov/easey/facilities-mgmt/facilities/attributes?page=", p, "&perPage=100&year=", y, sep="")
    facilities_attributes = request(url)%>%
      req_headers(`x-api-key`=API_Key,
                  `accept`= "application/json")%>%
      req_perform()
    facilities_attributes = fromJSON(rawToChar(facilities_attributes$body))
    facilities_attributes_data=rbind(facilities_attributes_data, facilities_attributes)
    print(p)
  }
}
write.csv(facilities_attributes_data, "Data/CAMD/facilities_attributes.csv")


facilities_emissions_data=NULL
for(y in 2015:2020){
  url=paste("https://api.epa.gov/easey/emissions-mgmt/emissions/apportioned/annual?page=1&perPage=100&year=", y, sep="")
  facilities_emissions = request(url)%>%
    req_headers(`x-api-key`=API_Key,
                `accept`= "application/json")%>%
    req_perform()
  
  num_pages = ceiling(as.numeric(facilities_emissions$headers$`X-Total-Count`)/100)
  
  facilities_emissions = fromJSON(rawToChar(facilities_emissions$body))
  facilities_emissions_data=rbind(facilities_emissions_data, facilities_emissions)
  for(p in 2:num_pages){
    url=paste("https://api.epa.gov/easey/emissions-mgmt/emissions/apportioned/annual?page=", p, "&perPage=100&year=", y, sep="")
    facilities_emissions = request(url)%>%
      req_headers(`x-api-key`=API_Key,
                  `accept`= "application/json")%>%
      req_perform()
    facilities_emissions = fromJSON(rawToChar(facilities_emissions$body))
    facilities_emissions_data=rbind(facilities_emissions_data, facilities_emissions)
    print(p)
  }
}
write.csv(facilities_emissions_data, "Data/CAMD/facilities_emissions.csv")

File: Get_CAMD.py
import requests
import json
import pandas as pd

# Set API key and base URL
# API Keys can be requested at https://www.epa.gov/power-sector/cam-api-portal#/api-key-signup
f = open("MIP_AirPollution/API_Key.json")
config = json.load(f) 
API_Key = config['API_Key']

data_directory = "C:/Users/lbeatty/Documents/Lauren_MIP_Contribution/Data/"
base_url = "https://api.epa.gov/easey/"

# Function to fetch data
def fetch_data(endpoint, year):
    data = []
    page = 1
    while True:
        url = f"{base_url}{endpoint}?page={page}&perPage=100&year={year}"
        headers = {
            "x-api-key": API_Key,
            "accept": "application/json"
        }
        response = requests.get(url, headers=headers)
        if response.status_code == 200:
            content = response.json()
            data.extend(content)
            total_count = int(response.headers.get('X-Total-Count', 0))
            if total_count <= page * 100:
                break
            page += 1
        else:
            print(f"Failed to fetch data for year {year}, page {page}")
            break
    return data

# Fetch facilities attributes data
facilities_attributes_data = []
for year in range(2015, 2021):
    endpoint = f"facilities-mgmt/facilities/attributes"
    facilities_attributes_data.extend(fetch_data(endpoint, year))

# Convert to DataFrame and save as CSV
facilities_attributes_df = pd.DataFrame(facilities_attributes_data)
facilities_attributes_df.to_csv(data_directory+"facilities_attributes.csv", index=False)

# Fetch facilities emissions data
facilities_emissions_data = []
for year in range(2015, 2021):
    endpoint = f"emissions-mgmt/emissions/apportioned/annual"
    facilities_emissions_data.extend(fetch_data(endpoint, year))

# Convert to DataFrame and save as CSV
facilities_emissions_df = pd.DataFrame(facilities_emissions_data)
facilities_emissions_df.to_csv(data_directory+"facilities_emissions.csv", index=False)